File: client-py/broker_client.py
"""
Broker Trading API Client for Python
"""

import hashlib
import hmac
import json
import time
import random
import base64
from typing import Optional, Dict, Any, List
import requests


class APIError(Exception):
    """API Error"""
    def __init__(self, code: str, message: str, request_id: str = ""):
        super().__init__(f"API Error [{code}]: {message} (RequestID: {request_id})")
        self.code = code
        self.message = message
        self.request_id = request_id


class Balance:
    """Asset balance"""
    def __init__(self, asset: str, total: str, locked: str):
        self.asset = asset
        self.total = total
        self.locked = locked

    def __repr__(self):
        return f"Balance(asset='{self.asset}', total='{self.total}', locked='{self.locked}')"


class RouteStep:
    """Swap route step"""
    def __init__(self, exchange: str, pool: str, from_asset: str, to_asset: str, amount_in: str, amount_out: str):
        self.exchange = exchange
        self.pool = pool
        self.from_asset = from_asset
        self.to_asset = to_asset
        self.amount_in = amount_in
        self.amount_out = amount_out

    def __repr__(self):
        return f"RouteStep(exchange='{self.exchange}', from_asset='{self.from_asset}', to_asset='{self.to_asset}')"


class BrokerClient:
    """Client for Broker Trading API"""

    def __init__(self, api_key: str, secret_key: str, base_url: str):
        """
        Initialize API client

        Args:
            api_key: API key
            secret_key: Secret key
            base_url: Base API URL
        """
        self.api_key = api_key
        self.secret_key = secret_key
        self.base_url = base_url
        self.session = requests.Session()
        self.session.timeout = 30

    def generate_nonce(self) -> str:
        """Generate unique nonce"""
        timestamp_ns = int(time.time() * 1_000_000_000)
        random_part = random.randint(0, 999_999)
        return f"{timestamp_ns}_{random_part}"

    def hash_body(self, body: str) -> str:
        """Calculate SHA256 hash of request body"""
        if not body:
            body = ""
        return hashlib.sha256(body.encode('utf-8')).hexdigest()

    def generate_signature(self, method: str, path_with_query: str, timestamp: int, nonce: str, body_sha256: str) -> str:
        """
        Generate HMAC-SHA256 signature

        Args:
            method: HTTP method
            path_with_query: Path with query parameters
            timestamp: Timestamp in milliseconds
            nonce: Nonce
            body_sha256: SHA256 hash of request body

        Returns:
            HMAC-SHA256 signature in hex format
        """
        canonical_string = "\n".join([
            method.upper(),
            path_with_query,
            str(timestamp),
            nonce,
            body_sha256
        ])

        # Debug information
        print("Creating signature:")
        print(f"Method: {method.upper()}")
        print(f"Path: {path_with_query}")
        print(f"Timestamp: {timestamp}")
        print(f"Nonce: {nonce}")
        print(f"BodySHA256: {body_sha256}")
        print(f"CanonicalString: {repr(canonical_string)}")

        # Hash secret key and encode in base64url
        secret_key_hash = hashlib.sha256(self.secret_key.encode('utf-8')).digest()
        secret_key_base64 = base64.urlsafe_b64encode(secret_key_hash).decode('utf-8')
        
        # Create HMAC signature
        signature = hmac.new(
            secret_key_base64.encode('utf-8'),
            canonical_string.encode('utf-8'),
            hashlib.sha256
        ).hexdigest()

        print(f"Signature: {signature}")
        print("---")

        return signature

    def make_request(self, method: str, path: str, body: Optional[Dict[str, Any]] = None, 
                     additional_headers: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
        """
        Make authenticated request to API

        Args:
            method: HTTP method
            path: API path
            body: Request body
            additional_headers: Additional headers

        Returns:
            API response as dictionary

        Raises:
            APIError: API error
            requests.RequestException: Network error
        """
        timestamp = int(time.time() * 1000)  # milliseconds
        nonce = self.generate_nonce()
        
        body_string = json.dumps(body, separators=(',', ':')) if body else ""
        body_sha256 = self.hash_body(body_string)

        signature = self.generate_signature(method, path, timestamp, nonce, body_sha256)

        headers = {
            'Content-Type': 'application/json',
            'X-API-KEY': self.api_key,
            'X-API-TIMESTAMP': str(timestamp),
            'X-API-NONCE': nonce,
            'X-API-SIGN': signature,
        }

        if additional_headers:
            headers.update(additional_headers)

        url = self.base_url + path
        
        response = self.session.request(
            method=method,
            url=url,
            headers=headers,
            data=body_string if body_string else None
        )

        response.raise_for_status()

        # Parse response
        try:
            response_data = response.json()
        except json.JSONDecodeError:
            raise APIError("PARSE_ERROR", f"Failed to parse response: {response.text}")

        # Check if response is an API error
        if isinstance(response_data, dict) and 'code' in response_data and 'message' in response_data:
            raise APIError(
                code=response_data['code'],
                message=response_data['message'],
                request_id=response_data.get('requestId', '')
            )

        return response_data

    def get_balances(self) -> List[Balance]:
        """
        Get user balances

        Returns:
            List of asset balances
        """
        response = self.make_request('GET', '/api/v1/balances')
        balances_data = response.get('balances', [])
        return [Balance(
            asset=b['asset'],
            total=b['total'],
            locked=b['locked']
        ) for b in balances_data]

    def estimate_swap(self, from_asset: str, to_asset: str, amount: str, 
                      network: Optional[str] = None, account: Optional[str] = None) -> Dict[str, Any]:
        """
        Get swap estimation

        Args:
            from_asset: Source asset
            to_asset: Target asset
            amount: Amount
            network: Network (optional)
            account: Account (optional)

        Returns:
            Estimation data
        """
        request_data = {
            'from': from_asset,
            'to': to_asset,
            'amount': amount
        }
        
        if network:
            request_data['network'] = network
        if account:
            request_data['account'] = account

        return self.make_request('POST', '/api/v1/estimate', request_data)

    def swap(self, from_asset: str, to_asset: str, amount: str, account: str,
             slippage_bps: int, idempotency_key: str, client_order_id: Optional[str] = None) -> Dict[str, Any]:
        """
        Execute swap

        Args:
            from_asset: Source asset
            to_asset: Target asset
            amount: Amount
            account: Account
            slippage_bps: Slippage in basis points
            idempotency_key: Idempotency key
            client_order_id: Client order ID (optional)

        Returns:
            Created swap data
        """
        request_data = {
            'from': from_asset,
            'to': to_asset,
            'amount': amount,
            'account': account,
            'slippage_bps': slippage_bps
        }
        
        if client_order_id:
            request_data['clientOrderId'] = client_order_id

        headers = {'Idempotency-Key': idempotency_key}
        
        return self.make_request('POST', '/api/v1/swap', request_data, headers)

    def get_order_status(self, order_id: str, client_order_id: Optional[str] = None) -> Dict[str, Any]:
        """
        Get order status

        Args:
            order_id: Order ID
            client_order_id: Client order ID (optional)

        Returns:
            Order status
        """
        path = f'/api/v1/orders/{order_id}/status'
        if client_order_id:
            path += f'?clientOrderId={client_order_id}'

        return self.make_request('GET', path)

File: client-js/broker-client.js
import crypto from 'crypto';
import fetch from 'node-fetch';

/**
 * TheOne Trading API Client for JavaScript
 */
export class BrokerClient {
    /**
     * @param {string} apiKey - API key
     * @param {string} secretKey - Secret key
     * @param {string} baseURL - Base API URL
     */
    constructor(apiKey, secretKey, baseURL) {
        this.apiKey = apiKey;
        this.secretKey = secretKey;
        this.baseURL = baseURL;
    }

    /**
     * Generates unique nonce
     * @returns {string}
     */
    generateNonce() {
        const timestamp = Date.now() * 1000000; // nanoseconds
        const random = Math.floor(Math.random() * 1000000);
        return `${timestamp}_${random}`;
    }

    /**
     * Calculates SHA256 hash of request body
     * @param {string|null} body - Request body
     * @returns {string}
     */
    hashBody(body) {
        if (!body) {
            body = '';
        }
        return crypto.createHash('sha256').update(body).digest('hex');
    }

    /**
     * Generates HMAC-SHA256 signature
     * @param {string} method - HTTP method
     * @param {string} pathWithQuery - Path with query parameters
     * @param {number} timestamp - Timestamp in milliseconds
     * @param {string} nonce - Nonce
     * @param {string} bodySHA256 - SHA256 hash of request body
     * @returns {string}
     */
    generateSignature(method, pathWithQuery, timestamp, nonce, bodySHA256) {
        const canonicalString = [
            method.toUpperCase(),
            pathWithQuery,
            timestamp.toString(),
            nonce,
            bodySHA256
        ].join('\n');

        // Hash the secret key with SHA256
        const secretKeyHash = crypto.createHash('sha256').update(this.secretKey).digest();
        // Encode as base64 (not base64url) to match Go implementation that includes padding
        const secretKeyBase64 = Buffer.from(secretKeyHash).toString('base64')
            .replace(/\+/g, '-')
            .replace(/\//g, '_');
        // In Go: []byte(base64.URLEncoding.EncodeToString(hash[:])) - this means bytes of the string, not decode
        const hmacKey = Buffer.from(secretKeyBase64, 'utf8');
        
        const signature = crypto
            .createHmac('sha256', hmacKey)
            .update(canonicalString)
            .digest('hex');

        return signature;
    }

    /**
     * Makes authenticated request to API
     * @param {string} method - HTTP method
     * @param {string} path - API path
     * @param {object|null} body - Request body
     * @param {object} additionalHeaders - Additional headers
     * @returns {Promise<object>}
     */
    async makeRequest(method, path, body = null, additionalHeaders = {}) {
        const timestamp = Date.now();
        const nonce = this.generateNonce();
        
        const bodyString = body ? JSON.stringify(body) : '';
        const bodySHA256 = this.hashBody(bodyString);

        const signature = this.generateSignature(method, path, timestamp, nonce, bodySHA256);

        const headers = {
            'Content-Type': 'application/json',
            'X-API-KEY': this.apiKey,
            'X-API-TIMESTAMP': timestamp.toString(),
            'X-API-NONCE': nonce,
            'X-API-SIGN': signature,
            ...additionalHeaders
        };

        const url = this.baseURL + path;
        const options = {
            method,
            headers,
            body: bodyString || undefined
        };

        const response = await fetch(url, options);
        const responseText = await response.text();

        if (!response.ok) {
            throw new Error(`HTTP error ${response.status}: ${responseText}`);
        }

        // Check if response is an API error
        let responseData;
        try {
            responseData = JSON.parse(responseText);
            if (responseData.code && responseData.message) {
                const error = new Error(`API Error [${responseData.code}]: ${responseData.message}`);
                error.code = responseData.code;
                error.requestId = responseData.requestId;
                throw error;
            }
        } catch (parseError) {
            if (parseError.code) {
                throw parseError; // This is an API error, re-throw
            }
            throw new Error(`Failed to parse response: ${parseError.message}`);
        }

        return responseData;
    }

    /**
     * Gets user balances
     * @returns {Promise<object>}
     */
    async getBalances() {
        return await this.makeRequest('GET', '/api/v1/balances');
    }

    /**
     * Gets swap estimation
     * @param {object} request - Estimation parameters
     * @param {string} request.from - Source asset
     * @param {string} request.to - Target asset
     * @param {string} request.amount - Amount
     * @param {string} [request.network] - Network
     * @param {string} [request.account] - Account
     * @returns {Promise<object>}
     */
    async estimateSwap(request) {
        return await this.makeRequest('POST', '/api/v1/estimate', request);
    }

    /**
     * Executes swap
     * @param {object} request - Swap parameters
     * @param {string} request.from - Source asset
     * @param {string} request.to - Target asset
     * @param {string} request.amount - Amount
     * @param {string} request.account - Account
     * @param {number} request.slippage_bps - Slippage in basis points
     * @param {string} [request.clientOrderId] - Client order ID
     * @param {string} idempotencyKey - Idempotency key
     * @returns {Promise<object>}
     */
    async swap(request, idempotencyKey) {
        return await this.makeRequest('POST', '/api/v1/swap', request, {
            'Idempotency-Key': idempotencyKey
        });
    }

    /**
     * Gets order status
     * @param {string} orderId - Order ID
     * @param {string} [clientOrderId] - Client order ID
     * @returns {Promise<object>}
     */
    async getOrderStatus(orderId, clientOrderId = null) {
        let path = `/api/v1/orders/${orderId}/status`;
        if (clientOrderId) {
            path += `?clientOrderId=${encodeURIComponent(clientOrderId)}`;
        }
        return await this.makeRequest('GET', path);
    }
}
